File: src/main/java/org/edtp/universe/level/UniverseBundle.java
package org.edtp.universe.level;

import net.casual.arcade.dimensions.level.CustomLevel;
import org.edtp.universe.model.UniverseDimension;

import java.util.EnumMap;
import java.util.Objects;
import java.util.UUID;

public record UniverseBundle(
    UUID owner,
    long generation,
    EnumMap<UniverseDimension, CustomLevel> levels
) {
    public UniverseBundle {
        Objects.requireNonNull(owner, "owner");
        Objects.requireNonNull(levels, "levels");
    }

    public CustomLevel get(UniverseDimension dimension) {
        return Objects.requireNonNull(
            levels.get(dimension),
            () -> "Missing " + dimension + " in universe " + owner + " generation " + generation
        );
    }
}

File: src/main/kotlin/org/edtp/universe/level/UniverseBundle.kt
package org.edtp.universe.level

import net.casual.arcade.dimensions.level.CustomLevel
import org.edtp.universe.model.UniverseDimension
import java.util.EnumMap
import java.util.UUID

data class UniverseBundle(
    val owner: UUID,
    val generation: Long,
    val levels: EnumMap<UniverseDimension, CustomLevel>,
) {
    operator fun get(dimension: UniverseDimension): CustomLevel =
        requireNotNull(levels[dimension]) { "Missing $dimension in universe $owner generation $generation" }
}
